// @tempots/std/arrays.js
/*
Copyright 2019 Google LLC
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    https://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
/**
 * Utility functions to manipulate `Array` values.
 */
import { Maybe } from './maybe';
import { keys } from './objects';
export function map(arr, f) {
    return Array.from({ length: arr.length }, (_, i) => f(arr[i], i));
}
export function mapNotNull(arr, f) {
    const buff = [];
    for (let i = 0; i < arr.length; i++) {
        const v = f(arr[i], i);
        if (v != null) {
            buff.push(v);
        }
    }
    return buff;
}
export function flatMap(arr, f) {
    const buff = [];
    for (const el of arr) {
        buff.push(...f(el));
    }
    return buff;
}
export function head(arr) {
    return arr.length > 0 ? arr[0] : Maybe.nothing;
}
export function tail(arr) {
    return arr.slice(1);
}
export function equals(a, b, equality) {
    if (a.length !== b.length)
        return false;
    else {
        for (let i = 0; i < a.length; i++) {
            if (!equality(a[i], b[i]))
                return false;
        }
        return true;
    }
}
export function makeEquals(equality) {
    return function (a, b) {
        return equals(a, b, equality);
    };
}
export function isEmpty(arr) {
    return arr.length === 0;
}
export function hasValues(arr) {
    return arr.length > 0;
}
export function filter(arr, predicate) {
    const buff = [];
    for (const a of arr)
        if (predicate(a))
            buff.push(a);
    return buff;
}
export function filterNulls(arr) {
    return filter(arr, v => v != null);
}
export function flatten(arr) {
    return [].concat(...arr);
}
export function foldLeft(arr, f, b) {
    for (const a of arr) {
        b = f(b, a);
    }
    return b;
}
export function all(arr, predicate) {
    for (const a of arr) {
        if (!predicate(a)) {
            return false;
        }
    }
    return true;
}
export function any(arr, predicate) {
    for (const a of arr) {
        if (predicate(a)) {
            return true;
        }
    }
    return false;
}
export function each(arr, f) {
    for (const a of arr)
        f(a);
}
export function concat(...arrs) {
    return [].concat(...arrs);
}
export function makeCompare(comparef, shorterFirst = true) {
    return function (a, b) {
        if (a.length < b.length) {
            return -1 * (shorterFirst ? 1 : -1);
        }
        else if (a.length > b.length) {
            return 1 * (shorterFirst ? 1 : -1);
        }
        for (let i = 0; i < a.length; i++) {
            const ord = comparef(a[i], b[i]);
            if (ord !== 0)
                return ord;
        }
        return 0;
    };
}
export function sort(compare, arr) {
    return arr.slice().sort(compare);
}
export function range(length, f) {
    return Array.from({ length }, (_, i) => f(i));
}
export function numbersRange(length, startAt = 0) {
    return range(length, i => startAt + i);
}
export function fill(length, value) {
    return range(length, () => value);
}
export function distinctPrimitive(values) {
    return Array.from(new Set(values));
}
export function distinctByPredicate(values, predicate) {
    const map = {};
    values.forEach(v => {
        map[predicate(v)] = v;
    });
    return keys(map).map(k => map[k]);
}
export function remove(arr, item, predicate) {
    let index;
    if (predicate !== undefined) {
        index = arr.findIndex(predicate);
    }
    else {
        index = arr.indexOf(item);
    }
    if (index < 0) {
        return false;
    }
    else {
        arr.splice(index, 1);
        return true;
    }
}
export function ofIterableIterator(it) {
    const buff = [];
    for (let r = it.next(); !(r.done ?? false); r = it.next()) {
        buff.push(r.value);
    }
    return buff;
}
export function diffOperations(from, to, getKey) {
    const ops = {
        removals: [],
        swaps: [],
        inserts: []
    };
    const { removals, inserts, swaps } = ops;
    const mapB = new Map();
    to.forEach((v, i) => mapB.set(getKey(v), i));
    const indexesOfAThatDoNotExistInB = from
        .map((v, i) => [v, i])
        .filter(([v]) => !mapB.has(getKey(v)))
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        .map(([_, i]) => i);
    for (let i = indexesOfAThatDoNotExistInB.length - 1; i >= 0; i--) {
        const p = indexesOfAThatDoNotExistInB[i];
        const last = removals.length > 0 ? removals[removals.length - 1] : undefined;
        if ((last != null) && last.at === p + 1) {
            last.at--;
            last.qt++;
        }
        else {
            removals.push({ at: p, qt: 1 });
        }
    }
    const mapA = new Map();
    from.forEach((v, i) => mapA.set(getKey(v), i));
    const indexesOfBThatDoNotExistInA = to
        .map((v, i) => [v, i])
        .filter(([v]) => !mapA.has(getKey(v)))
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        .map(([_, i]) => i);
    for (const p of indexesOfBThatDoNotExistInA) {
        const last = inserts.length > 0 ? inserts[inserts.length - 1] : undefined;
        if ((last != null) && last.at + last.values.length === p) {
            last.values.push(to[p]);
        }
        else {
            inserts.push({ at: p, values: [to[p]] });
        }
    }
    const ra = from.filter((_, i) => !indexesOfAThatDoNotExistInB.includes(i));
    const mapRA = new Map();
    for (let i = 0; i < ra.length; i++) {
        mapRA.set(getKey(ra[i]), i);
    }
    const rb = to.filter((_, i) => !indexesOfBThatDoNotExistInA.includes(i));
    for (let i = 0; i < rb.length; i++) {
        const bk = getKey(rb[i]);
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const ai = mapRA.get(bk);
        if (ai == null || i === ai)
            continue;
        const ak = getKey(ra[i]);
        mapRA.delete(ak);
        swaps.push({ from: i, to: ai });
    }
    return ops;
}
export function applyOperations(operations, start) {
    const buff = [...start];
    for (const { at, qt } of operations.removals) {
        buff.splice(at, qt);
    }
    for (const { from, to } of operations.swaps) {
        const t = buff[to];
        buff[to] = buff[from];
        buff[from] = t;
    }
    for (const op of operations.inserts) {
        buff.splice(op.at, 0, ...op.values);
    }
    return buff;
}
export function joinWithConjunction(arr, conjunction = ' and ', separator = ', ') {
    if (arr.length === 0)
        return '';
    if (arr.length === 1)
        return String(arr[0]);
    return `${arr.slice(0, -1).join(separator)}${conjunction}${String(arr[arr.length - 1])}`;
}
