// @tempots/std/functions.js
/*
Copyright 2019 Google LLC
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    https://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function compose(...args) {
    return (a) => args.reduce((acc, f) => f(acc), a);
}
export function identity(v) {
    return v;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function curryLeft(f) {
    return (a) => (...rest) => f(a, ...rest);
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function curryRight(f) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (...args) => 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (a) => f(...args.concat([a]));
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function flip(f) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (...args) => f(...args.reverse());
}
export function id(v) {
    return v;
}
