// @tempots/std/match.js
/*
Copyright 2019 Google LLC
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    https://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
export function matchLiteral(input, matcher) {
    return matcher[input];
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function match(input, field, matcher) {
    const k = input[field];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return matcher[k](input);
}
export function deepMatch(input, path, matcher) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const k = path.reduce((res, key) => res[key], input);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return matcher[k](input);
}
export function createMatch(field) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (input, matcher) => {
        const k = input[field];
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return matcher[k](input);
    };
}
export const createDeepMatch = (...path) => 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
(input, matcher) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const k = path.reduce((res, key) => res[key], input);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return matcher[k](input);
};
export const matchKind = createMatch('kind');
export const matchType = createMatch('type');
