// @tempots/std/ord.js
/*
Copyright 2019 Google LLC
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    https://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
export var Ordering;
(function (Ordering) {
    Ordering[Ordering["LT"] = -1] = "LT";
    Ordering[Ordering["EQ"] = 0] = "EQ";
    Ordering[Ordering["GT"] = 1] = "GT";
})(Ordering || (Ordering = {}));
export class Ord {
    compare;
    static fromNumberComparison(compare) {
        return new Ord(fromNumberComparison(compare));
    }
    constructor(compare) {
        this.compare = compare;
    }
    max(a, b) {
        switch (this.compare(a, b)) {
            case Ordering.LT:
            case Ordering.EQ:
                return a;
            case Ordering.GT:
                return b;
        }
    }
    min(a, b) {
        switch (this.compare(a, b)) {
            case Ordering.GT:
            case Ordering.EQ:
                return a;
            case Ordering.LT:
                return b;
        }
    }
    equals(a, b) {
        switch (this.compare(a, b)) {
            case Ordering.EQ:
                return true;
            default:
                return false;
        }
    }
    contramap(f) {
        return new Ord((b0, b1) => this.compare(f(b0), f(b1)));
    }
    inverse() {
        return new Ord((a0, a1) => this.compare(a1, a0));
    }
    numberComparison(a0, a1) {
        return this.compare(a0, a1);
    }
}
export function fromNumberComparison(f) {
    return function compare(a, b) {
        const r = f(a, b);
        if (r < 0)
            return Ordering.LT;
        else if (r === 0)
            return Ordering.EQ;
        else
            return Ordering.GT;
    };
}
