// happy-dom/lib/css/CSSRuleTypeEnum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CSSRuleTypeEnum;
(function (CSSRuleTypeEnum) {
    CSSRuleTypeEnum[CSSRuleTypeEnum["containerRule"] = 0] = "containerRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["styleRule"] = 1] = "styleRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["importRule"] = 3] = "importRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["mediaRule"] = 4] = "mediaRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["fontFaceRule"] = 5] = "fontFaceRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["pageRule"] = 6] = "pageRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["keyframesRule"] = 7] = "keyframesRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["keyframeRule"] = 8] = "keyframeRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["namespaceRule"] = 10] = "namespaceRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["counterStyleRule"] = 11] = "counterStyleRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["supportsRule"] = 12] = "supportsRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["documentRule"] = 13] = "documentRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["fontFeatureValuesRule"] = 14] = "fontFeatureValuesRule";
    CSSRuleTypeEnum[CSSRuleTypeEnum["regionStyleRule"] = 16] = "regionStyleRule";
})(CSSRuleTypeEnum || (CSSRuleTypeEnum = {}));
exports.default = CSSRuleTypeEnum;
