// @tempots/ui/components/Tooltip/index.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "@tempots/dom/jsx-runtime";
import { Consumer, Signal, Attribute } from '@tempots/dom';
import { UnstyledPopover } from '../Popover';
import { onTargetOverMount } from '../Popover/UnstyledPopover';
import { StyleMarker } from '../StyleProvider/StyleProvider';
import { Sx } from '../styling/Sx';
const ARROW_SIZE = 8;
export function Tooltip({ children }) {
    return (_jsx(_Fragment, { children: _jsx(Consumer, { mark: StyleMarker, children: ({ styles }) => {
                const at = styles.at('styles').at;
                return (_jsxs(UnstyledPopover, { onTargetMount: onTargetOverMount, placement: Signal.of('top'), arrow: _jsx("div", { children: _jsx(Sx, { sx: {
                                backgroundColor: at('background').at('inverseColor'),
                                boxShadow: at('shadow').at('md'),
                                marginTop: -ARROW_SIZE / 2,
                                transform: 'rotate(45deg)',
                                width: ARROW_SIZE,
                                height: ARROW_SIZE
                            } }) }), children: [_jsx(Attribute, { name: "role", value: Signal.of('tooltip') }), _jsx(Sx, { sx: {
                                maxWidth: '200px',
                                height: 'auto',
                                lineHeight: '1em',
                                textAlign: 'center',
                                whiteSpace: 'normal',
                                backgroundColor: at('background').at('inverseColor'),
                                color: at('font').at('inverseColor'),
                                padding: at('spacing').map(v => `${v.xs}px ${v.md}px`),
                                borderRadius: at('border').at('radius').at('sm'),
                                boxShadow: at('shadow').at('md')
                            } }), children] }));
            } }) }));
}
