// happy-dom/lib/css/declaration/utilities/CSSStyleDeclarationElementStyle.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const CSSStyleDeclarationPropertyManager_1 = __importDefault(require("./CSSStyleDeclarationPropertyManager"));
const NodeTypeEnum_1 = __importDefault(require("../../../nodes/node/NodeTypeEnum"));
const CSSRuleTypeEnum_1 = __importDefault(require("../../CSSRuleTypeEnum"));
const CSSStyleDeclarationElementDefaultCSS_1 = __importDefault(require("./CSSStyleDeclarationElementDefaultCSS"));
const CSSStyleDeclarationElementInheritedProperties_1 = __importDefault(require("./CSSStyleDeclarationElementInheritedProperties"));
const CSSStyleDeclarationCSSParser_1 = __importDefault(require("./CSSStyleDeclarationCSSParser"));
const QuerySelector_1 = __importDefault(require("../../../query-selector/QuerySelector"));
const CSS_VARIABLE_REGEXP = /var\( *(--[^) ]+)\)/g;
/**
 * CSS Style Declaration utility
 */
class CSSStyleDeclarationElementStyle {
    /**
     * Constructor.
     *
     * @param element Element.
     * @param [computed] Computed.
     */
    constructor(element, computed = false) {
        this.cache = {
            propertyManager: null,
            cssText: null,
            documentCacheID: null
        };
        this.element = element;
        this.computed = computed;
    }
    /**
     * Returns element style properties.
     *
     * @returns Element style properties.
     */
    getElementStyle() {
        if (this.computed) {
            return this.getComputedElementStyle();
        }
        const cssText = this.element['_attributes']['style']?.value;
        if (cssText) {
            if (this.cache.propertyManager && this.cache.cssText === cssText) {
                return this.cache.propertyManager;
            }
            this.cache.cssText = cssText;
            this.cache.propertyManager = new CSSStyleDeclarationPropertyManager_1.default({ cssText });
            return this.cache.propertyManager;
        }
        return new CSSStyleDeclarationPropertyManager_1.default();
    }
    /**
     * Returns style sheets.
     *
     * @param element Element.
     * @returns Style sheets.
     */
    getComputedElementStyle() {
        const documentElements = [];
        const parentElements = [];
        let styleAndElement = {
            element: this.element,
            cssTexts: []
        };
        let shadowRootElements = [];
        if (!this.element.isConnected) {
            return new CSSStyleDeclarationPropertyManager_1.default();
        }
        if (this.cache.propertyManager &&
            this.cache.documentCacheID === this.element.ownerDocument['_cacheID']) {
            return this.cache.propertyManager;
        }
        this.cache.documentCacheID = this.element.ownerDocument['_cacheID'];
        // Walks through all parent elements and stores them in an array with element and matching CSS text.
        while (styleAndElement.element) {
            if (styleAndElement.element.nodeType === NodeTypeEnum_1.default.elementNode) {
                const rootNode = styleAndElement.element.getRootNode();
                if (rootNode.nodeType === NodeTypeEnum_1.default.documentNode) {
                    documentElements.unshift(styleAndElement);
                }
                else {
                    shadowRootElements.unshift(styleAndElement);
                }
                parentElements.unshift(styleAndElement);
            }
            if (styleAndElement.element === this.element.ownerDocument) {
                const styleSheets = (this.element.ownerDocument.querySelectorAll('style,link[rel="stylesheet"]'));
                for (const styleSheet of styleSheets) {
                    const sheet = styleSheet.sheet;
                    if (sheet) {
                        this.parseCSSRules({
                            elements: documentElements,
                            cssRules: sheet.cssRules
                        });
                    }
                }
                styleAndElement = { element: null, cssTexts: [] };
            }
            else if (styleAndElement.element.nodeType === NodeTypeEnum_1.default.documentFragmentNode &&
                styleAndElement.element.host) {
                const styleSheets = (styleAndElement.element.querySelectorAll('style,link[rel="stylesheet"]'));
                styleAndElement = {
                    element: styleAndElement.element.host,
                    cssTexts: []
                };
                for (const styleSheet of styleSheets) {
                    const sheet = styleSheet.sheet;
                    if (sheet) {
                        this.parseCSSRules({
                            elements: shadowRootElements,
                            cssRules: sheet.cssRules,
                            hostElement: styleAndElement
                        });
                    }
                }
                shadowRootElements = [];
            }
            else {
                styleAndElement = { element: styleAndElement.element.parentNode, cssTexts: [] };
            }
        }
        // Concatenates all parent element CSS to one string.
        const targetElement = parentElements[parentElements.length - 1];
        let inheritedCSSText = '';
        for (const parentElement of parentElements) {
            if (parentElement !== targetElement) {
                parentElement.cssTexts.sort((a, b) => a.priorityWeight - b.priorityWeight);
                if (CSSStyleDeclarationElementDefaultCSS_1.default[parentElement.element.tagName]) {
                    inheritedCSSText +=
                        CSSStyleDeclarationElementDefaultCSS_1.default[parentElement.element.tagName];
                }
                for (const cssText of parentElement.cssTexts) {
                    inheritedCSSText += cssText.cssText;
                }
                if (parentElement.element['_attributes']['style']?.value) {
                    inheritedCSSText += parentElement.element['_attributes']['style'].value;
                }
            }
        }
        const cssVariables = {};
        const properties = {};
        let targetCSSText = CSSStyleDeclarationElementDefaultCSS_1.default[targetElement.element.tagName] || '';
        targetElement.cssTexts.sort((a, b) => a.priorityWeight - b.priorityWeight);
        for (const cssText of targetElement.cssTexts) {
            targetCSSText += cssText.cssText;
        }
        if (targetElement.element['_attributes']['style']?.value) {
            targetCSSText += targetElement.element['_attributes']['style'].value;
        }
        const combinedCSSText = inheritedCSSText + targetCSSText;
        if (this.cache.propertyManager && this.cache.cssText === combinedCSSText) {
            return this.cache.propertyManager;
        }
        // Parses the parent element CSS and stores CSS variables and inherited properties.
        CSSStyleDeclarationCSSParser_1.default.parse(inheritedCSSText, (name, value, important) => {
            if (name.startsWith('--')) {
                const cssValue = this.getCSSValue(value, cssVariables);
                if (cssValue) {
                    cssVariables[name] = cssValue;
                }
                return;
            }
            if (CSSStyleDeclarationElementInheritedProperties_1.default[name]) {
                const cssValue = this.getCSSValue(value, cssVariables);
                if (cssValue && (!properties[name]?.important || important)) {
                    properties[name] = {
                        value: cssValue,
                        important
                    };
                }
            }
        });
        // Parses the target element CSS.
        CSSStyleDeclarationCSSParser_1.default.parse(targetCSSText, (name, value, important) => {
            if (name.startsWith('--')) {
                const cssValue = this.getCSSValue(value, cssVariables);
                if (cssValue && (!properties[name]?.important || important)) {
                    cssVariables[name] = cssValue;
                    properties[name] = {
                        value,
                        important
                    };
                }
            }
            else {
                const cssValue = this.getCSSValue(value, cssVariables);
                if (cssValue && (!properties[name]?.important || important)) {
                    properties[name] = {
                        value: cssValue,
                        important
                    };
                }
            }
        });
        const propertyManager = new CSSStyleDeclarationPropertyManager_1.default();
        for (const name of Object.keys(properties)) {
            propertyManager.set(name, properties[name].value, properties[name].important);
        }
        this.cache.cssText = combinedCSSText;
        this.cache.propertyManager = propertyManager;
        return propertyManager;
    }
    /**
     * Applies CSS text to elements.
     *
     * @param options Options.
     * @param options.elements Elements.
     * @param options.cssRules CSS rules.
     * @param [options.hostElement] Host element.
     */
    parseCSSRules(options) {
        if (!options.elements.length) {
            return;
        }
        const defaultView = options.elements[0].element.ownerDocument.defaultView;
        for (const rule of options.cssRules) {
            if (rule.type === CSSRuleTypeEnum_1.default.styleRule) {
                const selectorText = rule.selectorText;
                if (selectorText) {
                    if (selectorText.startsWith(':host')) {
                        if (options.hostElement) {
                            options.hostElement.cssTexts.push({
                                cssText: rule._cssText,
                                priorityWeight: 0
                            });
                        }
                    }
                    else {
                        for (const element of options.elements) {
                            const matchResult = QuerySelector_1.default.match(element.element, selectorText);
                            if (matchResult.matches) {
                                element.cssTexts.push({
                                    cssText: rule._cssText,
                                    priorityWeight: matchResult.priorityWeight
                                });
                            }
                        }
                    }
                }
            }
            else if (rule.type === CSSRuleTypeEnum_1.default.mediaRule &&
                defaultView.matchMedia(rule.conditionText).matches) {
                this.parseCSSRules({
                    elements: options.elements,
                    cssRules: rule.cssRules,
                    hostElement: options.hostElement
                });
            }
        }
    }
    /**
     * Returns CSS value.
     *
     * @param value Value.
     * @param cssVariables CSS variables.
     * @returns CSS value.
     */
    getCSSValue(value, cssVariables) {
        const regexp = new RegExp(CSS_VARIABLE_REGEXP);
        let newValue = value;
        let match;
        while ((match = regexp.exec(value)) !== null) {
            const cssVariableValue = cssVariables[match[1]];
            if (!cssVariableValue) {
                return null;
            }
            newValue = newValue.replace(match[0], cssVariableValue);
        }
        return newValue;
    }
}
exports.default = CSSStyleDeclarationElementStyle;
