// dicerollerts/roll-result.js
export function oneResult(die) {
    return {
        type: 'one-result',
        die
    };
}
export function literalResult(value, result) {
    return {
        type: 'literal-result',
        value,
        result
    };
}
export function diceReduceResult(reduceables, reducer, result) {
    return {
        type: 'dice-reduce-result',
        reduceables,
        reducer,
        result
    };
}
export function binaryOpResult(op, left, right, result) {
    return {
        type: 'binary-op-result',
        op,
        left,
        right,
        result
    };
}
export function unaryOpResult(op, expr, result) {
    return {
        type: 'unary-op-result',
        op,
        expr,
        result
    };
}
export function diceExpressionsResult(rolls) {
    return {
        type: 'dice-expressions-result',
        rolls
    };
}
export function diceFilterableResult(rolls, filter) {
    return {
        type: 'dice-filterable-result',
        rolls,
        filter
    };
}
export function diceMapeableResult(rolls, functor) {
    return {
        type: 'dice-mapeable-result',
        rolls,
        functor
    };
}
export function rerolled(rerolls) {
    return {
        type: 'rerolled',
        rerolls
    };
}
export function exploded(explosions) {
    return {
        type: 'exploded',
        explosions
    };
}
export function normal(roll) {
    return {
        type: 'normal',
        roll
    };
}
export function dieResult(result, sides) {
    return {
        type: 'die-result',
        result,
        sides
    };
}
export function keepResult(roll) {
    return {
        type: 'keep-result',
        roll
    };
}
export function discardResult(roll) {
    return {
        type: 'discard-result',
        roll
    };
}
